src/envoy/auth/config.cc

```diff
@@ -15,199 +15,119 @@
-  ENVOY_LOG(debug, "AsyncClientCallbacks [cluster = {}]: failed",
-            cluster_->name());
-  cb_(false, "");
+  return "";
 }
 
-void AsyncClientCallbacks::Call(const std::string &uri) {
-  ENVOY_LOG(debug, "AsyncClientCallbacks [cluster = {}]: {} {}",
-            cluster_->name(), __func__, uri);
-  // Example:
-  // uri  = "https://example.com/certs"
-  // pos  :          ^
-  // pos1 :                     ^
-  // host = "example.com"
-  // path = "/certs"
-  auto pos = uri.find("://");
-  pos = pos == std::string::npos ? 0 : pos + 3;  // Start position of host
-  auto pos1 = uri.find("/", pos);
-  if (pos1 == std::string::npos) pos1 = uri.length();
-  std::string host = uri.substr(pos, pos1 - pos);
-  std::string path = "/" + uri.substr(pos1 + 1);
-
-  MessagePtr message(new RequestMessageImpl());
-  message->headers().insertMethod().value().setReference(
-      Http::Headers::get().MethodValues.Get);
-  message->headers().insertPath().value(path);
-  message->headers().insertHost().value(host);
-
-  request_ = cm_.httpAsyncClientForCluster(cluster_->name())
-                 .send(std::move(message), *this, timeout_);
+Config::Config(std::vector<IssuerInfo>&& issuers)
+    : issuers_(std::move(issuers)) {
+  ENVOY_LOG(debug, "Config: {}", __func__);
 }
 
-void AsyncClientCallbacks::Cancel() { request_->cancel(); }
+Config::Config(const Json::Object& config) {
+  ENVOY_LOG(debug, "Config: {}", __func__);
 
-IssuerInfo::IssuerInfo(Json::Object *json) {
-  ENVOY_LOG(debug, "IssuerInfo: {}", __func__);
-  // Check "name"
-  name_ = json->getString("name", "");
-  if (name_ == "") {
-    ENVOY_LOG(debug, "IssuerInfo: Issuer name missing");
-    failed_ = true;
+  // Load the issuers as JSON array.
+  std::vector<Json::ObjectSharedPtr> issuer_jsons;
+  try {
+    issuer_jsons = config.getObjectArray("issuers");
+  } catch (...) {
+    ENVOY_LOG(error, "Config: issuers should be array type");
     return;
   }
-  // Check "audience". It will be an empty array if the key "audience" does not
-  // exist
+
+  for (auto issuer_json : issuer_jsons) {
+    IssuerInfo issuer;
+    if (LoadIssuerInfo(*issuer_json, &issuer)) {
+      std::string err = issuer.Validate();
+      if (err.empty()) {
+        issuers_.push_back(issuer);
+      } else {
+        ENVOY_LOG(error, "Issuer [name = {}]: {}", issuer.name, err);
+      }
+    }
+  }
+}
+
+bool Config::LoadIssuerInfo(const Json::Object& json, IssuerInfo* issuer) {
+  // Check "name"
+  issuer->name = json.getString("name", "");
+  if (issuer->name == "") {
+    ENVOY_LOG(error, "Issuer: Issuer name missing");
+    return false;
+  }
+
+  // Check "audience".
+  // It will be an empty array if the key "audience" does not exist.
   try {
-    audiences_ = json->getStringArray("audiences", true);
+    auto audiences = json.getStringArray("audiences", true);
+    issuer->audiences.insert(audiences.begin(), audiences.end());
   } catch (...) {
```

Comment:
It may be better for debug (as well as readability) to catch and print out the exception message. Eg.
catch (const Exception& e) {
ENVOY_LOG(error, e.what());
}

Just my personal opinion, feel free to ignore.